include/boost/beast/experimental/core/uri/detail/parser.hpp

```diff
@@ -0,0 +1,469 @@
+//
+// Copyright (c) 2018 oneiric (oneiric at i2pmail dot org)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/beast
+//
+
+#ifndef BOOST_BEAST_CORE_URI_DETAIL_PARSER_HPP
+#define BOOST_BEAST_CORE_URI_DETAIL_PARSER_HPP
+
+#include <boost/beast/core/string.hpp>
+#include <boost/beast/experimental/core/uri/buffer.hpp>
+#include <boost/beast/experimental/core/uri/rfc3986.hpp>
+
+namespace boost {
+namespace beast {
+namespace uri {
+namespace detail {
+
+class parser_impl {
+  template <typename It>
+  It find_delimiter_or_mismatch(It first, It last,
+                                std::function<bool(char)> delimiter_func,
+                                std::function<bool(char)> match_func,
```

Comment:
Using std::function here seems like a premature pessimization. Have you considered taking the callback functor types as template parameters instead?